design/thrift-proxy-design.md

```diff
@@ -0,0 +1,215 @@
+Users can define routing rules specific to their application's needs, ensuring that Thrift traffic is efficiently directed to the appropriate services within the cluster.
+```yaml
+apiVersion: projectcontour.io/v1
+kind: ThriftProxy
+metadata:
+  name: thriftproxy-1
+  namespace: projectcontour
+spec:
+  routes:
+    - condition:
+        methodName: <method_name>
+        headers:
+          - name: <some_header>
+            exact: <value>
+      services:
+        - name: <cluster_name>
+          port: <port_number>
+```
+
+### Example
+```yaml
+apiVersion: projectcontour.io/v1
+kind: ThriftProxy
+metadata:
+  name: placeorder
+  namespace: placeorder-ns
+spec:
+  routes:
+    - condition:
+        methodName: "placeOrder"
+        headers:
+          - name: "Authorization"
+            exact: "Bearer token"
+      services:
+        - name: "order-service"
+          port: 9090
+
+```
+
+## Thrift Listener
+When a new ThriftProxy resource is discovered by Contour, a corresponding Thrift Listener is automatically added to the Envoy configuration. 
+This listener is configured with default settings to ensure seamless integration into Contour.
+### Default Settings:
+* Name: The Thrift Listener is named "ingress_thrift" for identification within the Envoy configuration.
+* Protocol: The protocol is set to "thrift" to differentiate Thrift traffic from other protocols.
+* Address and Port: The address and port are configurable, allowing users to specify the desired listening address and port number, similar to how the HTTP listener operates (default: :9090)
+* RouteConfigName: The Thrift Listener is associated with a default Route Configuration named "ingress_thrift.".
+
+The Thrift Listener includes a Thrift filter with essential configurations:
+```
+&envoy_listener_v3.Filter{
+    StatPrefix: name,
+    AccessLog:  accesslogger,
+    Trds: &envoy_thrift_v3.Trds{
+        RouteConfigName: name,
+        ConfigSource: &envoy_core_v3.ConfigSource{
+            ResourceApiVersion: envoy_core_v3.ApiVersion_V3,
+            ConfigSourceSpecifier: &envoy_core_v3.ConfigSource_Ads{
+                Ads: &envoy_core_v3.AggregatedConfigSource{},
+            },
+        },
+    },
+}
+```
+
+## Thrift RouteConfiguration
+A Thrift RouteConfiguration is essentially an organized collection of Thrift routes.
+Thrift routes are defined as follows:
+```
+&envoy_thrift_v3.Route{
+    Match: ThriftRouteMatch(rule),
+    Route: ThriftRouteAction(rule),
+}
+```
+* ThriftRouteMatch: Defines the matching rules that incoming Thrift traffic must satisfy to be routed according to this rule. These rules can be based on factors such as the Thrift method name and custom headers.
+* ThriftRouteAction: Specifies the cluster that should handle the Thrift traffic that matches the given rule. This cluster configuration includes details such as load balancing policies, fail-over strategies, and rate limiting.
+
+## Aggregated Service Discovery
```

Comment:
Huh interesting that TRDS is limited to ADS basically, might be worth adding a link to the field where it says that: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto#extensions-filters-network-thrift-proxy-v3-trds (we can't use the plain gRPC SOTW api config source)